# Homework 02 — Decimal-to-Binary Conversion and IEEE 754

## Problem

Implement, using **pure Python**, two functions to convert decimal numbers to their binary representation. The use of NumPy and built-in conversion functions such as `bin()` is not allowed.

### Required Functions

**1. Integer-to-binary conversion**

Implement the function

```python
entero_a_binario(n)
```

using the **successive division by 2 algorithm**.

**2. Fraction-to-binary conversion**

Implement the function

```python
fraccion_a_binario(f, max_bits=20)
```

using the **successive multiplication by 2 algorithm**. If the binary representation does not terminate, the algorithm must stop after `max_bits` binary digits.

As a reference,

$$
(11.1875)_{10} = (1011.0011)_2.
$$

## Tasks

**a)** Apply the implemented functions to obtain the binary representations of

$$
18.0308, \qquad 26.0985, \qquad 94.0516.
$$

Use the reference conversion above to verify the implementation.

In [ ]:
def entero_a_binario(n):
  if n == 0:
    return "0"
  binario = ""
  while n > 0:
    residuo = n % 2
    binario = str(residuo) + binario
    n = n // 2
  return binario

def fraccion_a_binario(n, max_bits = 20):
  binario_decimal = ""
  binario_entero = entero_a_binario(int(n//1))
  parte_decimal = n - (n//1)
  contador = 0
  while parte_decimal != 0 and contador < max_bits:
    parte_decimal = parte_decimal * 2
    if parte_decimal > 1:
      binario_decimal += "1"
      parte_decimal = parte_decimal - 1
    elif parte_decimal==1:
      binario_decimal += "1"
      break
    else:
      binario_decimal += "0"
    contador += 1
  return str(binario_entero)+"."+str(binario_decimal)

In [ ]:
#Comprobaciones propias
numero1 = entero_a_binario(13)
print(f"El numero binario es: {numero1}")
x = fraccion_a_binario(0.725)
print(x)

El numero binario es: 1101
0.10111001100110011001


In [ ]:
#Punto a
#Primero verifiquemos con el ejemplo del profesor, nos dice que {11.1875}_10 = {1011.0011}_2
print(f"(11.1875)_10:({fraccion_a_binario(11.1875)})_2")

(11.1875)_10:(1011.0011)_2


In [ ]:
#Continuación Punto a
# Ahora que ya vimos que al parecer esta bien el codigo, hacemos las conversiones que nos piden
print(f"(18.0308)_10:({fraccion_a_binario(18.0308)})_2")
print(f"(26.0985)_10:({fraccion_a_binario(26.0985)})_2")
print(f"(94.0516)_10:({fraccion_a_binario(94.0516)})_2")

(18.0308)_10:(10010.00000111111000101000)_2
(26.0985)_10:(11010.00011001001101110100)_2
(94.0516)_10:(1011110.00001101001101011010)_2


**b)** Determine the binary representation of

$$
0.3
$$

using `max_bits = 20`. Notice that its binary representation does not terminate exactly.

In [ ]:
#Punto b
print(f"(0.3)_10:({fraccion_a_binario(0.3)})_2")
#0.3 tiene representación binaria infinita

(0.3)_10:(0.01001100110011001100)_2



**c)** Evaluate $0.1$ and $0.2$ and explain why, in Python,

```python
0.1 + 0.2 != 0.3
```

Relate this behavior to the finite binary representation of floating-point numbers.

In [ ]:
#Punto c
print(f"(0.1)_10:({fraccion_a_binario(0.1)})_2")
print(f"(0.2)_10:({fraccion_a_binario(0.2)})_2")
print(f"(0.3)_10:({fraccion_a_binario(0.3)})_2")
print(f"(0.4)_10:({fraccion_a_binario(0.4)})_2")
print(f"(0.5)_10:({fraccion_a_binario(0.5)})_2")

(0.1)_10:(0.00011001100110011001)_2
(0.2)_10:(0.00110011001100110011)_2
(0.3)_10:(0.01001100110011001100)_2
(0.4)_10:(0.01100110011001100110)_2
(0.5)_10:(0.1)_2


In [ ]:
# Continuacion c
0.1+0.2

0.30000000000000004

In [ ]:
# Continuacion c
#Vemos que 1,2,3 y 4 tienen representacion binaria infinita periodica, sin embargo, esta no es respuesta
# satisfactoria para explicar por que da distinto de 3 pues vemos que por ejemplo:
print(0.1+0.3)

0.4


## Continuacion c:
Entonces, realmente que hace que no sea exactamente 3? son 2 factores, primero si tiene que ver que
tenga representacion binaria infinita, pero tambien importa que los pequeños errores de aproximación
de los sumandos hagan que el resultado quede cerca del float que represente el resultado de la suma.

En estos casos es: El error de aproximacion de 0.1 + 0.2 hacia que combinados, la suma quedara mas cerca
de 0.30000000000000004 que del float que representa 0.3, en cambio para 0.1 y 0.3 los errores se
combinan de manera tal que el resultado coincide con el mismo float que representa 0.4.

**d)** — IEEE 754 Float32 Encoding

Using the functions developed in class in the previous parts of the exercise, implement a function in **pure Python** that follows the **IEEE 754 encoding algorithm for 32-bit floating-point numbers (`float32`)**.

Apply the function to the same decimal numbers considered in part **(a)**:

$$
18.0308, \qquad 26.0985, \qquad 94.0516.
$$

For each number, obtain its IEEE 754 `float32` representation and compare the results with the reference functions presented in class.

In [ ]:
def ieee754_float32(n):

    # Bit de signo
    if n < 0:
        signo = "1"
        n = -n
    else:
        signo = "0"

    numero_binario = fraccion_a_binario(n, max_bits=50)
    binario_entero, binario_decimal = numero_binario.split(".")
    exponente = len(binario_entero) - 1
    mantisa_completa = binario_entero[1:] + binario_decimal

    while len(mantisa_completa) < 24:
        mantisa_completa += "0"

    mantisa = mantisa_completa[:23]
    bit_redondeo = mantisa_completa[23]

    if bit_redondeo == "1":

        mantisa_entero = 0

        for bit in mantisa:
            mantisa_entero = mantisa_entero * 2 + int(bit)

        mantisa_entero += 1

        if mantisa_entero == 2**23:
            mantisa = "0" * 23
            exponente += 1
        else:
            mantisa = entero_a_binario(mantisa_entero)

            while len(mantisa) < 23:
                mantisa = "0" + mantisa

    exponente_bias = exponente + 127
    exponente_binario = entero_a_binario(exponente_bias)

    while len(exponente_binario) < 8:
        exponente_binario = "0" + exponente_binario

    resultado = signo + exponente_binario + mantisa
    return resultado

In [ ]:
# Punto d
numeros = [18.0308, 26.0985, 94.0516]

for numero in numeros:

    resultado = ieee754_float32(numero)

    print("Número:", numero)
    print("Signo:", resultado[0])
    print("Exponente:", resultado[1:9])
    print("Mantisa:", resultado[9:])
    print("IEEE 754 float32:", resultado)
    print()

Número: 18.0308
Signo: 0
Exponente: 10000011
Mantisa: 00100000011111100010100
IEEE 754 float32: 01000001100100000011111100010100

Número: 26.0985
Signo: 0
Exponente: 10000011
Mantisa: 10100001100100110111010
IEEE 754 float32: 01000001110100001100100110111010

Número: 94.0516
Signo: 0
Exponente: 10000101
Mantisa: 01111000001101001101011
IEEE 754 float32: 01000010101111000001101001101011

